### ISSUE TITLE
Calva does not gracefully handle when clojure-lsp cannot be downloaded

### ISSUE THREAD

**Issue by jacobmolin** [BetterThanTomorrow/calva#2064]
I'm running Calva (version 2.0.331) in WSL2 on a Windows 10 machine. I have nothing in Clojure Lsp Path and Clojure Lsp Verison is "latest". I have never gotten clojure-lsp to work on WSL and I've been using Calva on WSL since early January 2023.

When I open/start a clojure project in VSCode I get:
![clojure-lsp-1](https://user-images.githubusercontent.com/34372356/217023716-2513af58-2f4c-4aff-932b-d4c904e0332b.png)

The output:
```
[Error - 4:51:02 PM] Clojure Language Client client: couldn't create connection to server.
Launching server using command /home/sejamol/.vscode-server/extensions/betterthantomorrow.calva-2.0.331/clojure-lsp failed. Error: spawn /home/sejamol/.vscode-server/extensions/betterthantomorrow.calva-2.0.331/clojure-lsp ENOENT
```

The clojure-lsp is not active, so I press "clojure-lsp" in the status bar and then "Download latest clojure-lsp version". This gives me the following messages (the path in the message is "/home/sejamol/.vscode-server/extensions/betterthantomorrow.calva-2.0.331/clojure-lsp":
![clojure-lsp-2](https://user-images.githubusercontent.com/34372356/217023855-dcec450c-74e7-47e6-bdfc-e16fc4b1527e.png)

However, there's no "clojure-lsp" in ".../betterthantomorrow.calva-2.0.331/". When I reload VSCode or try to start the clojure-lsp the same thing happens as described above.

So I'm guessing there's an issue with downloading the clojure-lsp. I've read through #1590, but haven't found a solution for me yet.

**Comment by bpringe** [comment]
Thanks for reporting! There were some changes around this lately. You can probably work around this issue for now by installing version 2.0.326 of Calva.

@julienvincent, could you take a look at this?

CC @PEZ

**Comment by jacobmolin** [comment]
I installed version 2.0.326 of Calva but I still get the error "Clojure Language Client client: couldn't create connection to server." Also, there's no longer a "clojure-lsp" in the status bar, only "REPL" and "clj".

**Comment by PEZ** [comment]
Yes, I don't think this is due to the clojure-lsp updates. Can you see in the VS Code development console log any messages that could give a clue? (The Help menu has an item for opening the console.)

**Comment by jacobmolin** [comment]
A correction, after a reboot, I can see the clojure-lsp in the status bar, but it disappears after unsuccessfully trying to start/download it.

Here's the whole development console content after startup of a clojure project. There seems to be some issue with a proxy (2nd image). I'm using a work computer, but I don't have a proxy on when trying this.

![developer-log-1](https://user-images.githubusercontent.com/34372356/217278827-23e3cf7a-741d-45ab-8c71-4bd302a9bdb4.png)
![developer-log-2](https://user-images.githubusercontent.com/34372356/217278832-e62a5645-06b7-445d-b8d5-05f7a390aec3.png)

I don't know if it matters, but I've had issues accessing servers running in WSL via localhost in the host. I've solved this issue by binding the WSL IP address to localhost in the host each time WSL is restarted. The clojure-lsp issue was present before this localhost ordeal though.

**Comment by PEZ** [comment]
Can you download things using `wget` from the terminal? Specifically that clojure-lsp zip file. Both external and VS Code internal terminal is of interest. I think the 407 error might mean that there is a proxy somewhere in the network. But I have to admit I am not very knowledgable about things like this.

**Comment by jacobmolin** [comment]
I can download things using `wget`, both in the VS Code internal terminal and external, but not the clojure-lsp zip file. I get 404:
```
$ wget <https://github.com/clojure-lsp/clojure-lsp/releases/download//clojure-lsp-native-static-linux-amd64.zip>
--2023-02-09 09:27:07--  <https://github.com/clojure-lsp/clojure-lsp/releases/download//clojure-lsp-native-static-linux-amd64.zip>
Resolving github.com (github.com)... 140.82.121.4
Connecting to github.com (github.com)|140.82.121.4|:443... connected.
HTTP request sent, awaiting response... 404 Not Found
2023-02-09 09:27:07 ERROR 404: Not Found.
```

I took the URL from the 2nd image in my previous comment (around the middle). I think it looks a little weird with the double "/" after "download". When I go to the release page manually in the browser and copy the download link I get:
```
<https://github.com/clojure-lsp/clojure-lsp/releases/download/2023.01.26-11.08.16/clojure-lsp-native-linux-amd64.zip>
```

This URL contains the release tag after the "download/". Maybe the previous URL is missing the tag?

I also tried putting the clojure-lsp-file in the betterthantomorrow extension path after downloading the zip manually. I get the same error but the developer logno longer complain about the clojure-lsp download but some other download . I don't know if it's related to clojure-lsp.
![developer-log-3](https://user-images.githubusercontent.com/34372356/217768008-39d5ed59-8c7f-433c-ab77-4ca1e4c015ab.png)

**Comment by bpringe** [comment]
> This URL contains the release tag after the "download/". Maybe the previous URL is missing the tag?

Yes, that seems like the issue. There's probably an issue with however Calva is getting that release tag.

**Comment by bpringe** [comment]
@jacobmolin What's your `calva.clojureLspVersion` setting set to?

**Comment by bpringe** [comment]
I'm guessing that `downloadVersion` is probably `''` here: https://github.com/BetterThanTomorrow/calva/blob/bdfcd416d2494cb165c01c02d45b26b2e64c2ee2/src/lsp/client/downloader.ts#L182

due to the below function returning `''` because of an exception:

```typescript
async function getLatestVersion(): Promise<string> {
  try {
    const releasesJSON = await util.fetchFromUrl(
      'https://api.github.com/repos/clojure-lsp/clojure-lsp/releases'
    );
    const releases = JSON.parse(releasesJSON);
    return releases[0].tag_name;
  } catch (err) {
    return '';
  }
}
```

@jacobmolin Can you run Calva in development mode and see if that's what's happening, and see if you can figure out why?

**Comment by bpringe** [comment]
I'm guessing maybe this logic in your case is allowing `downloadClojureLsp` to be called with `''` as the `downloadVersion`:

```typescript
  const exists = await fs.promises
    .stat(clojureLspPath)
    .then(() => true)
    .catch((err) => {
      if (err.code !== 'ENOENT') {
        throw err;
      }
      return false;
    });

  if (
    (currentVersion !== downloadVersion && downloadVersion !== '') ||
    forceDownload ||
    downloadVersion === 'nightly' ||
    !exists
  ) {
    await downloadClojureLsp(context.extensionPath, downloadVersion);
  }
```

Maybe `exists` is false, which I think would lead to what I said.

**Comment by jacobmolin** [comment]
Sorry about the late reply!

> @jacobmolin Can you run Calva in development mode and see if that's what's happening, and see if you can figure out why?

I think I could check it out, but I'm not sure how to run Calva in development mode. Is it to follow the [How-to-Hack-on-Calva](https://github.com/BetterThanTomorrow/calva/wiki/How-to-Hack-on-Calva) guide?

**Comment by bpringe** [comment]
Yes, that should get you up and running in development mode.

**Comment by jacobmolin** [comment]
> I'm guessing that `downloadVersion` is probably `''` here:
> 
> https://github.com/BetterThanTomorrow/calva/blob/bdfcd416d2494cb165c01c02d45b26b2e64c2ee2/src/lsp/client/downloader.ts#L182
> 
> due to the below function returning `''` because of an exception:
> 
> ```ts
> async function getLatestVersion(): Promise<string> {
>   try {
>     const releasesJSON = await util.fetchFromUrl(
>       'https://api.github.com/repos/clojure-lsp/clojure-lsp/releases'
>     );
>     const releases = JSON.parse(releasesJSON);
>     return releases[0].tag_name;
>   } catch (err) {
>     return '';
>   }
> }
> ```
> 
@bpringe Correct! The error is `Error [ERR_SOCKET_CLOSED]: Socket is closed`. I tried using the URL with `curl` and `wget` in the terminal of both WSL and Windows and there's no problem getting the 200 response.

> I'm guessing maybe this logic in your case is allowing `downloadClojureLsp` to be called with `''` as the `downloadVersion`:
> 
> ```ts
>   const exists = await fs.promises
>     .stat(clojureLspPath)
>     .then(() => true)
>     .catch((err) => {
>       if (err.code !== 'ENOENT') {
>         throw err;
>       }
>       return false;
>     });
> 
>   if (
>     (currentVersion !== downloadVersion && downloadVersion !== '') ||
>     forceDownload ||
>     downloadVersion === 'nightly' ||
>     !exists
>   ) {
>     await downloadClojureLsp(context.extensionPath, downloadVersion);
>   }
> ```
> 
> Maybe `exists` is false, which I think would lead to what I said.

@bpringe Also correct.

**Comment by jacobmolin** [comment]
I stumbled upon [this issue](https://github.com/microsoft/vscode/issues/115818) in VSCode and tried the[ suggestion of setting the “Http: Proxy Support” to off](https://github.com/microsoft/vscode/issues/115818#issuecomment-774898877) and that solved it. I don't know the implications of turning proxy support off, but at least Calva now downloads clojure-lsp and it runs when I jack-in. I’m happy to provide more info if you’re interested.

**Comment by bpringe** [comment]
@jacobmolin Thanks for confirming the above! Glad you found a workaround. We can probably fix this in Calva so you don't need to use that workaround. I need look into it a bit.

**Comment by bpringe** [comment]
@PEZ What do you think Calva should do in this situation (when clojure-lsp can't be downloaded and there's no path for a local clojure-lsp set)?

I'm thinking maybe we should show some message to the user to say that clojure-lsp could not be downloaded, maybe show the download error, and instruct them to download it themselves and set the path setting as an alternative route.

**Comment by PEZ** [comment]
Guiding the user to install their own clojure-lsp binary makes most sense.

**Comment by bpringe** [comment]
@jacobmolin Would you mind installing this vsix, turning your proxy support back on, reloading VS Code, and then telling us if you then see an error explaining that clojure-lsp couldn't be downloaded? If you don't see an error and you see clojure-lsp start, you may have to manually remove the clojure-lsp and the clojure-lsp-version files from your Calva extension directory so that Calva tries to download it again.

https://output.circle-artifacts.com/output/job/2054a1ff-0d16-4459-bb20-8988c17c7690/artifacts/0/tmp/artifacts/calva-2.0.336-fix-lsp-download-issue-f09aced5.vsix

@PEZ Would we consider that PR just addressing this issue, or fixing it? I'm not sure if there's something we could do to make the download work in this situation.

**Comment by PEZ** [comment]
Yes, it's much better this way.
